cmake/BuildDependencies.cmake

```diff
@@ -0,0 +1,171 @@
+macro(get_compiler_id)
+  # Get lowercase compiler id
+  string(TOLOWER "${CMAKE_CXX_COMPILER_ID}" cmake_cxx_compiler_id)
+  if(cmake_cxx_compiler_id STREQUAL "gnu")
+    set(cmake_cxx_compiler_id "gcc")
+  endif()
+endmacro()
+
+# if set toolchain file, don't build, and add include path and link path
+if(CMAKE_TOOLCHAIN_FILE)
+  set(BUILD_DEPENDENCIES
+      OFF
+      CACHE BOOL "Build Dependencies" FORCE
+  )
+  # toolchain file added include path and link path, use it
+  target_include_directories(${PROJECT_NAME} PRIVATE ${CMAKE_INCLUDE_PATH})
+  target_link_directories(${PROJECT_NAME} PRIVATE ${CMAKE_LIBRARY_PATH})
+endif()
+
+if(BUILD_DEPENDENCIES)
+  message(STATUS "Building Dependencies")
+
+  # CMake >= 3.14
+  if(${CMAKE_VERSION} VERSION_LESS "3.14")
+    message(FATAL_ERROR "CMake >= 3.14 required")
+  endif()
+
+  include(FetchContent)
+  # show progress
+  set(FETCHCONTENT_QUIET OFF)
+  set(FETCHCONTENT_UPDATES_DISCONNECTED ON)
+
+  # disable CMP0135 warning
+  if(${CMAKE_VERSION} VERSION_GREATER_EQUAL "3.24")
+    cmake_policy(SET CMP0135 NEW)
+  endif()
+
+  # 💡Using URL is faster than git clone
+
+  # GTest/GoogleTest
+  if(BUILD_TESTING)
+    #[[
+    # Can't use name ⚠️GTest/gtest⚠️ in FetchContent_Declare, see:
+    # CMake FetchContent not working as expected (https://github.com/google/googletest/issues/2457)
+    # FetchContent_* does not work with GTest as dependency name on Windows (https://gitlab.kitware.com/cmake/cmake/-/issues/25294)
+    # The name gtest reserved:
+    # googletest_SOURCE_DIR: build/_deps/googletest-src
+    # gtest_SOURCE_DIR: build/_deps/googletest-src/googletest
+    #]]
+    FetchContent_Declare(googletest URL "https://github.com/google/googletest/archive/refs/tags/v1.14.0.tar.gz")
+    # For Windows: Prevent overriding the parent project's compiler/linker ⚠️MT/MD⚠️ settings
+    set(gtest_force_shared_crt
+        ON
+        CACHE BOOL "" FORCE
+    )
+    FetchContent_MakeAvailable(googletest)
+
+    # Set build output path, this is mainly for Windows, to solve the test/unittests running 0x000135 problems
+    set_standard_build_output(gtest)
+    set_standard_build_output(gtest_main)
+    set_standard_build_output(gmock)
+    set_standard_build_output(gmock_main)
+  endif()
+
+  # spdlog
+  if(TRANTOR_USE_SPDLOG)
+    FetchContent_Declare(fmt URL "https://github.com/fmtlib/fmt/archive/refs/tags/10.2.1.tar.gz")
+    FetchContent_Declare(spdlog URL "https://github.com/gabime/spdlog/archive/refs/tags/v1.13.0.tar.gz")
+    FetchContent_MakeAvailable(fmt spdlog)
```

Comment:
I've questions regarding how are we going to maintain these. I have a feeling that without a need, these automatic dependency management will get out of sync with what's available and we won't have the main/brain power to maintain these.

Reply:
This is mainly for windows, and who don't want to install conan and vcpkg(it not easy to install it in some areas, such china mainland).

Do you know a way to download the latest version? 
Indeed, fetchcontent support GIT_REPOSITORY also, it could keep to latest.
```
FetchContent_Declare(
  googletest
  GIT_REPOSITORY https://github.com/google/googletest.git
  GIT_TAG        703bd9caab50b139428cea1aaff9974ebee5742e # release-1.10.0
)
```
but using GIT might cause git clone interupt issue due to the network problem. especially the big repository, such as openssl.

and using conan needs maintain the libs version also.

Reply:
Trantor does not depend on the fmt library. 
and if we fetch the source code of the famous libraries and build them into trantor, this can lead some version conflicts with the same libraries in OS used by users.

Reply:
spdlog needs fmt libray. if using conan or vcpkg, it will download it and build it automaticly.

Reply:
> Trantor does not depend on the fmt library. and if we fetch the source code of the famous libraries and build them into trantor, this can lead some version conflicts with the same libraries in OS used by users.

I think in Linux/Macos, it not nessesary to using BUILD_DEPENDENCIES(FetchContent) at all.  so BUILD_DEPENDENCIES default value is OFF.
But in windows, fmt is missing in most machines. And if you find the verison conflicts, you can modify the download url to keep version same.

Reply:
> Why don't we use the denpendencies mantained by vcpkg or connan.

Succesfully install vcpkg is not so easy due to git clone interupt. 
conan works fine, It just provide a third way.

Reply:
Users can use vcpkg or canon to manager dependencies on windows.
for fmt, you could install it by vcpkg (or conan), for example:
```shell
.\vcpkg.exe install fmt:x64-windows
```

Reply:
> Users can use vcpkg or canon to manager dependencies on windows. for fmt, you could install it by vcpkg (or conan), for example:
> 
> ```shell
> .\vcpkg.exe install fmt:x64-windows
> ```

I know it. if vcpkg installed successfully.

Reply:
If don't want to use CMAKE FectchContent as an option BUILD_DEPENDENCIES, I can delete these code.